test/gtest/ucs/test_event_set.cc

```diff
@@ -0,0 +1,140 @@
+/**
+* Copyright (C) Hiroyuki Sato. 2019. ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#include <common/test.h>
+extern "C" {
+#include <ucs/sys/event_set.h>
+#include <sys/epoll.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+}
+
+class test_event_set : public ucs::test {
+};
+
+static const char* const UCS_EVENT_SET_TEST_STRING = "OpenUCX event_set test";
+
+UCS_TEST_F(test_event_set, event_set_create) {
+    ucs_sys_event_set_t *event_set = NULL;
+    ucs_status_t status;
+
+    status = ucs_event_set_create(&event_set);
+    EXPECT_EQ(UCS_OK,status);
+    EXPECT_TRUE(event_set != NULL);
+
+    ucs_event_set_cleanup(event_set);
+}
+
+static ucs_status_t event_set_handler(int nready, void *event)
+{
+    struct epoll_event *epoll_event = (struct epoll_event *)event;
+
+    /* 
+     * TODO:
+     * event->data.ptr set in ucs_event_set_add function.
+     * This test check whether this value pass to handler correctly.
+     */
+    EXPECT_TRUE(strcmp((const char *)epoll_event->data.ptr,
+                       (const char *)UCS_EVENT_SET_TEST_STRING) == 0);
+    EXPECT_EQ(1,nready);
+    return UCS_OK;
+}
+
+UCS_TEST_F(test_event_set, ucs_event_set_read) {
+    int pipefd[2];
+    pid_t cpid;
+
+    if (pipe(pipefd) == -1) {
+        perror("pipe");
+    }
+
+    cpid = fork();
+    if (cpid == -1) {
+        perror("fork");
+    }
+
+    if (cpid == 0) { /* Child */
+        int n;
+        close(pipefd[0]);
+        usleep(10000); /* 10ms */
+        n = write(pipefd[1],"test",4);
+        if (n == -1) {
+            perror("write");
+        }
+        sleep(1);
+        close(pipefd[1]);
+        exit(0);
+
+    } else { /* Parent */
+        ucs_sys_event_set_t *event_set = NULL;
+        ucs_status_t status;
+        int wstatus;
+
+        close(pipefd[1]);
+
+        status = ucs_event_set_create(&event_set);
+        EXPECT_EQ(UCS_OK,status);
+        EXPECT_TRUE(event_set != NULL);
+
+        status = ucs_event_set_add(event_set,pipefd[0],UCS_EVENT_SET_EVREAD,
+                                   (void *)UCS_EVENT_SET_TEST_STRING);
+        EXPECT_EQ(UCS_OK,status);
+
+        status = ucs_event_set_wait(event_set,20,event_set_handler);
+        EXPECT_EQ(UCS_OK,status);
+        ucs_event_set_cleanup(event_set);
+        close(pipefd[0]);
+
+        if (waitpid(cpid,&wstatus,0) < 0) {
+            perror("waitpid");
+        };
+    }
+}
+
+UCS_TEST_F(test_event_set, ucs_event_set_event_timeout) {
+    int pipefd[2];
+    pid_t cpid;
+
+    if (pipe(pipefd) == -1) {
+        perror("pipe");
+    }
+
+    cpid = fork();
```

Comment:
why do we need to fork()? can just use pipes in same process. if needed, better create a thread

Reply:
you could use pipes from single process (or single thread) just as you would from multiple processes (read/write etc.) if need to avoid blocking on read/write, could use the event set mechanism, or create separate thread with pthread_create()

Reply:
Use pthread for avoiding blocking `epoll_wait`.